Add tshock installer executable

SignatureBeef opened the pull request:
This aims to fix a lot of the "the console just closes" issues we've been seeing recently by adding a second executable, `./TShock.Installer[.exe]`, to the archive for those who do not have dotnet installed globally on their system.

The installer will download the dotnet 6.0.11 runtime to a local ./dotnet folder, it will then set the DOTNET_ROOT environment variable that ./TShock.Server will proceed to use to run the non-self contained version.

**Why not just make TShock.Server be self contained?** 
1. Because many people have expressed they do not want dotnet in every installation, especially when considering GSP's whom already have it installed globally.
2. MonoMod has had issues with jit hooks, at least on arm64 when in self-contained + single file modes. however, this can be explored further due to recent changes, or via the new reorg branch which may allow clrjit to be found easier (at least for custom setups) which was difficult in the current version. alternatively, single file mode should just be avoided at the expense of more ./bin redirections, but point 1 takes precedence so this is not really effective either way.

**Will TShock.Server be affected?** 
No, it remains the same, the Installer will just setup dotnet, and call the existing setup which remains a singlefile but not self-contained (TShock.Server[.exe]).

**Can TShock.Server not be a single file so TShock.Server.dll and others are extracted?** 
Yes, but out of scope for this PR (and mainly just a flag in CI if someone wanted to change that later)

**Why do you hardcode the urls?** 
Because we don't support .net7 yet, nor did i want to overcomplicate the downloading using APIs. Happy for someone to do this, and ive left the reference to dotnet-install.[ps1/sh] which can be used to do this for when security/patches are released, but for now, 6.0.11 is known to work.

**Can TShock.Installer[.exe] be called x instead?** 
Yes, i have no emotional attachment, i only cared about the actual functionality, so happy to rename, or someone can rename after merging if there is some preference later before a release/doco.

_note, i would suggest updating doco if this gets merged to make people aware of this, but if they read the docs they would know to install dotnet to begin with, so hopefully by naming it TShock.Installer, they will try it by nature, and thus will download dotnet for them..._

**Below are what the installer looks like on various OS's:** 

_note, ignore `TShock Installer 6.0.0.0` being printed, fixed in a recent commit to show 5.0.0_

Linux ARM64 / RPI
![image](https://user-images.githubusercontent.com/776327/202931747-aa38a184-4ae9-4dc8-940a-b673f6ad2d74.png)

OSX x64 Intel:
![image](https://user-images.githubusercontent.com/776327/202932499-862bc799-dc37-478a-bfe8-75bb8321dfb5.png)

Ubuntu x64 (ubuntu-22.10-live-server-amd64):
![VirtualBox_Ubuntu_21_11_2022_09_34_39](https://user-images.githubusercontent.com/776327/202932640-f25bcee8-cb08-4089-8c6d-21a7b95cc78e.png)

Windows 11 x64:
![Screenshot 2022-11-21 094503](https://user-images.githubusercontent.com/776327/202933174-3a2061d5-cf02-4e0a-8899-26f3ef024d07.png)


<!-- Warning: If you create a pull request and wish to remain anonymous, you are highly advised to use Tails (https://tails.boum.org/) or a fresh git environment. We will *not* be able to help with anonymization after your pull request has been created. -->

<!-- Warning: If you do not update the changelog, your pull request will be ignored and eventually closed, without comment, without any support, and without any opinion or interaction from the TShock team. This is your only warning. -->

SignatureBeef committed:
Add TShock.Installer to download dotnet runtime

This will include a new ./TShock.Installer executable that will be for users without the dotnet runtime. This program will download the dotnet runtime, extract it, and then run ./TShock.Server for them using the downloaded runtime.

Note: only tested on osx, likely a no-go for linux/windows until more testing occurs.

SignatureBeef committed:
Merge branch 'Pryaxis:general-devel' into tshock-installer

SignatureBeef committed:
Add architecture variants to tshock installer

SignatureBeef committed:
Remove TShock.Server.dll reference

SignatureBeef committed:
Fix TShock Installer version being written to console

v5 not v6...

drunderscore commented on TShockInstaller/Program.cs:
```diff
@@ -0,0 +1,93 @@
+Console.WriteLine("Using url: " + url);
+
+var filename = url.Split('/').Last();
+var is_targz = filename.EndsWith(".tar.gz");
```
A bit nitpicky, but it's a bit weird to see `snake_case` being used instead of `camelCase`, but also I don't believe it matters _that_ much.

drunderscore reviewed (commented):
This is really cool and a good solution to Microsoft's unfortunate approach (or lack there of) to "informing the end user".

Is it intended that `TShock.Installer.pdb` is included with the CI artifacts?

Unfortunately, this sorta breaks the `SIGINT` handler the server has... or at least, makes them work different than intended & advertised, and causes exceptions.

At world selection, sending `SIGINT` will kill the installer process, but not the server process. Pressing Ctrl + C afterwards will not do anything (which is expected to force shutdown), and the server will stay in the background. Inserting anything into the terminal will cause the following exception:
```
System.IO.IOException: Input/output error
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirectory, Func`2 errorRewriter)
   at System.IO.StdInReader.ReadKey()
   at System.IO.StdInReader.ReadLineCore(Boolean consumeKeys)
   at System.IO.StdInReader.ReadLine()
   at System.IO.SyncTextReader.ReadLine()
   at System.Console.ReadLine()
   at Terraria.Main.DedServ()
   at Terraria.Program.orig_RunGame()
```

If instead you select a world and let it load, sending `SIGINT` will begin shutting down properly, and eventually exit. However, like above, you will be unable to force shutdown (because you cannot send a second `SIGINT`), and inserting anything into the terminal whilst the server process is still alive will cause the same exception as above.

SignatureBeef committed:
Prevent installer from closing via ctrl+c while server remains open

This allows tshock to handle CancelKeyPress as per normal, instead of closing the parent process, leaving the second trigger to never be reached.
see https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/TShock.cs#L710

SignatureBeef committed:
Embed symbols in the installer to remove .pdb

SignatureBeef commented:
> This is really cool and a good solution to Microsoft's unfortunate approach (or lack there of) to "informing the end user".
> 
> Is it intended that `TShock.Installer.pdb` is included with the CI artifacts?
> 
> Unfortunately, this sorta breaks the `SIGINT` handler the server has... or at least, makes them work different than intended & advertised, and causes exceptions.
> 
> At world selection, sending `SIGINT` will kill the installer process, but not the server process. Pressing Ctrl + C afterwards will not do anything (which is expected to force shutdown), and the server will stay in the background. Inserting anything into the terminal will cause the following exception:
> 
> ```
> System.IO.IOException: Input/output error
>    at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
>    at Interop.CheckIo(Int64 result, String path, Boolean isDirectory, Func`2 errorRewriter)
>    at System.IO.StdInReader.ReadKey()
>    at System.IO.StdInReader.ReadLineCore(Boolean consumeKeys)
>    at System.IO.StdInReader.ReadLine()
>    at System.IO.SyncTextReader.ReadLine()
>    at System.Console.ReadLine()
>    at Terraria.Main.DedServ()
>    at Terraria.Program.orig_RunGame()
> ```
> 
> If instead you select a world and let it load, sending `SIGINT` will begin shutting down properly, and eventually exit. However, like above, you will be unable to force shutdown (because you cannot send a second `SIGINT`), and inserting anything into the terminal whilst the server process is still alive will cause the same exception as above.


The PDB should probably be embedded like the server does, 403677fad2c11d0a29230e19a3edf04e7d325b64 fixes that.

SIGINT / CTRL + C, may now be sorted as per: fbd82bbce8802b0b5a10f836410aeaa7de0c0ef1
On windows at least, it seems to allow TShock to handle the `Console.CancelKeyPress` event which should allow the sub process to exit as expected.

drunderscore reviewed (commented):
Yup, PDB is now embedded and the `SIGINT` handler works as expected.

In my testing of screwing with the `SIGINT` handler, I interrupted (and corrupted) the download of the .NET runtime archive, which made successive executions always throw with `ICSharpCode.SharpZipLib.SharpZipBaseException: Unexpected EOF`, until I deleted the .NET runtime archive.

Would it make sense to catch any exceptions thrown when unarchiving, and delete the downloaded file?

SignatureBeef committed:
Remove installers sdk archive when archive fails

This will allow the next run to redownload the file.

SignatureBeef commented:
> Yup, PDB is now embedded and the `SIGINT` handler works as expected.
> 
> In my testing of screwing with the `SIGINT` handler, I interrupted (and corrupted) the download of the .NET runtime archive, which made successive executions always throw with `ICSharpCode.SharpZipLib.SharpZipBaseException: Unexpected EOF`, until I deleted the .NET runtime archive.
> 
> Would it make sense to catch any exceptions thrown when unarchiving, and delete the downloaded file?

Yeh of course, i quickly done e561158699286042154e4b0f8feeac2dabba4853 which should now account for this.

Using a corrupt runtime archive:
![image](https://user-images.githubusercontent.com/776327/203709221-c3b218e4-1cd9-40f5-8eda-1a79c7a35296.png)

After the catch cleans it up, allows a new download:
![image](https://user-images.githubusercontent.com/776327/203709079-2b82f1c0-79dd-44db-8543-905d1ed6bc48.png)

drunderscore reviewed (approved):
Awesome, this works wonderfully!

SignatureBeef committed:
Merge branch 'Pryaxis:general-devel' into tshock-installer
